### PR TITLE
Add .NET8.0 targets to projects

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Adds .NET8.0 target to all projects.

### CODE REVIEW

File: build/Common.props [hunk 1]
```diff
@@ -16,11 +16,11 @@
...
-    <TargetFrameworksForLibrariesExtended>net6.0;netstandard2.1;netstandard2.0;net462</TargetFrameworksForLibrariesExtended>
-    <TargetFrameworksForAspNetCoreInstrumentation>net7.0;net6.0;netstandard2.1;netstandard2.0</TargetFrameworksForAspNetCoreInstrumentation>
-    <TargetFrameworksForGrpcNetClientInstrumentation>net6.0;netstandard2.1;netstandard2.0</TargetFrameworksForGrpcNetClientInstrumentation>
-    <TargetFrameworksForPrometheusAspNetCore>net6.0</TargetFrameworksForPrometheusAspNetCore>
+    <TargetFrameworksForLibraries>net8.0;net6.0;netstandard2.0;net462</TargetFrameworksForLibraries>
+    <TargetFrameworksForLibrariesExtended>net8.0;net6.0;netstandard2.1;netstandard2.0;net462</TargetFrameworksForLibrariesExtended>
+    <TargetFrameworksForAspNetCoreInstrumentation>net8.0;net7.0;net6.0;netstandard2.1;netstandard2.0</TargetFrameworksForAspNetCoreInstrumentation>
+    <TargetFrameworksForGrpcNetClientInstrumentation>net8.0;net6.0;netstandard2.1;netstandard2.0</TargetFrameworksForGrpcNetClientInstrumentation>
+    <TargetFrameworksForPrometheusAspNetCore>net8.0;net6.0</TargetFrameworksForPrometheusAspNetCore>
```
Comment (reyang) [inline build/Common.props:23]: Do we need net8 here?

Comment (reyang) [reply build/Common.props:23]: I guess the answer is yes from https://github.com/open-telemetry/opentelemetry-dotnet/blob/dc1f09d4b968f96955ccfeaa5b0f6f7e9989656a/test/OpenTelemetry.AotCompatibility.TestApp/OpenTelemetry.AotCompatibility.TestApp.csproj#L16-L32. @Yun-Ting would you check if this PR covers all AOT/trim compatible packages?

**Review by reyang — APPROVED** [review]

File: build/Common.props (hunk 1 above)
Comment (Yun-Ting) [reply build/Common.props:23]: Yes, the PR does.

**Review by Yun-Ting — APPROVED** [review]
LGTM! Thank you.

File: src/OpenTelemetry.Api/.publicApi/Experimental/net8.0/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -0,0 +1,80 @@
+abstract OpenTelemetry.Logs.Logger.EmitLog(in OpenTelemetry.Logs.LogRecordData data, in OpenTelemetry.Logs.LogRecordAttributeList attributes) -> void
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Api/.publicApi/Experimental/net8.0/PublicAPI.Unshipped.txt:1]: Across this PR all of this stuff should go into the **shipped** files. These APIs have all shipped we don't want to allow them to be modified for net8.0 anymore than we want to allow them to be modified for other targets 😄

When we go to do the public api review for 1.7.0 what we need to see in the "unshipped" files is new stuff. This is going to add too much noise.

**Comment by CodeBlanch** [comment]
I asked @Yun-Ting if this was required for AIT and she said: no. So my question is, why do we need net8.0 on all the targets?

**Comment by utpilla** [comment]
> I asked @Yun-Ting if this was required for AOT and she said: no. So my question is, why do we need net8.0 on all the targets?

This goes back to the discussion for #4595, which added a net6.0 target to every project even if a particular project didn't really need any API from .NET 6. I think back then we agreed that doing this comes with a few benefits: overall ease of maintenance, newer code analysis etc.

**Comment by reyang** [comment]
> I asked @Yun-Ting if this was required for AOT and she said: no. So my question is, why do we need net8.0 on all the targets?

I think the answer is YES now. @Yun-Ting to confirm.

**Comment by vishweshbankwar** [comment]
> > I asked @Yun-Ting if this was required for AOT and she said: no. So my question is, why do we need net8.0 on all the targets?
> 
> This goes back to the discussion for #4595, which added a net6.0 target to every project even if a particular project didn't really need any API from .NET 6. I think back then we agreed that doing this comes with a few benefits: overall ease of maintenance, newer code analysis etc.

+1 And eventually .net6.0 targets will be removed next year.

File: src/OpenTelemetry.Api/.publicApi/Experimental/net8.0/PublicAPI.Unshipped.txt (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Api/.publicApi/Experimental/net8.0/PublicAPI.Unshipped.txt:1]: Woops looks like it is correct... this is an experimental one. Sorry @vishweshbankwar!

**Comment by reyang** [comment]
> > > I asked @Yun-Ting if this was required for AOT and she said: no. So my question is, why do we need net8.0 on all the targets?
> > 
> > 
> > This goes back to the discussion for #4595, which added a net6.0 target to every project even if a particular project didn't really need any API from .NET 6. I think back then we agreed that doing this comes with a few benefits: overall ease of maintenance, newer code analysis etc.
> 
> +1 And eventually .net6.0 targets will be removed next year.

+1

**Comment by CodeBlanch** [comment]
> I think back then we agreed that doing this comes with a few benefits: overall ease of maintenance, newer code analysis etc.

* Overall ease of maintenance I disagree with. Each new target adds a lot of maintenance burden. Api files to maintain being the primary headache. When it comes time to remove `net6.0` we can add whatever new supported target we need then. That way we lose the api files for one framework and gain them for another (net zero). Right now we are gaining burden for unclear benefit (for me).

* Code analysis comes with the SDK, not the targets. By upgrading to .NET 8 SDK we have already gained that. PS: For most projects we're not currently running the analyzers 🤣

I think we should consider adding the targets if/when we actually need something.

**Review by CodeBlanch — APPROVED** [review]
I'm not totally convinced adding these targets is good, but the work LGTM 😄

**Comment by vitek-karas** [comment]
> Code analysis comes with the SDK, not the targets. By upgrading to .NET 8 SDK we have already gained that. PS: For most projects we're not currently running the analyzers 🤣

This actually is not the case for trimming/AOT analyzer. Those are tied to the TFM of the project. (This is intentional because there's some interaction between the analyzers and the framework methods being annotated and maintaining a working matrix would be a giant headache).

That said, I would not consider this the sole reason to upgrade (unless you have more specific needs which are met by the upgrade). There will be some slight differences in analyzer and Native AOT compiler behavior anyway and in case of discrepancies the NativeAOT compiler is always the correct one (presuming no bugs in the compiler). It should be enough to run the AOT test app (in CI and locally) targeting `net8.0` as that will get you the .NET 8 NativeAOT compiler (the compiler also versions with the TFM of the app, not the SDK).

File: src/OpenTelemetry.Exporter.Prometheus.AspNetCore/PrometheusExporterMiddleware.cs [hunk 3]
```diff
@@ -70,7 +70,11 @@ public async Task InvokeAsync(HttpContext httpContext)
                 if (collectionResponse.View.Count > 0)
                 {
                     response.StatusCode = 200;
+#if NET8_0
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.Prometheus.AspNetCore/PrometheusExporterMiddleware.cs:73]: Shouldn't it be?

```suggestion
#if NET8_0_OR_GREATER
```

Comment (reyang) [reply src/OpenTelemetry.Exporter.Prometheus.AspNetCore/PrometheusExporterMiddleware.cs:73]: +1

Comment (vishweshbankwar) [reply src/OpenTelemetry.Exporter.Prometheus.AspNetCore/PrometheusExporterMiddleware.cs:73]: This was a typo. Fixed.

**Pull request merged**